# Euler problems 1 to 10.

I'll try to limit this to pure Python.

In [2]:
from IPython.core.debugger import set_trace

In [3]:
import math

## Problem 1
---
Multiples of 3 and 5

Problem 1 

If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.

Find the sum of all the multiples of 3 or 5 below 1000.

In [2]:
s = 0
for ix in range(1, 1000):
    if (ix%3==0) or (ix%5)==0:
        s += ix
        
print s

233168


## Problem 2
---
Even Fibonacci numbers

Problem 2 

Each new term in the Fibonacci sequence is generated by adding the previous two terms. By starting with 1 and 2, the first 10 terms will be:

1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ...

By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms.

In [3]:
s = 0
a, b = 1, 2
s += 2

LIMIT = 4000000

while (a<=LIMIT) and (b<=LIMIT):
    n = a + b
    if (n<=LIMIT) and (n%2==0):
        s += n
        
    a = b
    b = n
    
print s

4613732


## Problem 3
---
Largest prime factor

Problem 3 

The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?

In [4]:
primes = [2, 3, 5, 7, 11, 13]

In [5]:
import math

In [6]:
primes = [2]

number = 14
limit = int(math.floor(math.sqrt(number))) + 1
cur_largest = -1
for i in range(3, limit):
    if number%i==0:
        is_prime = True
        for p in primes:
            if i%p == 0:
                is_prime = False
                break

        if is_prime:
            primes.append(i)
            if i>cur_largest:
                cur_largest = i
        
print primes, cur_largest

[2] -1


I'm assuming that the input `number` is non-prime. Doing this allows me to iterate upto only `sqrt(number)`.

Unfortunately, this doesn't really work for small numbers.
A much direct (albiet naive) approach, would be to recursively factorize the number.
1. start with 2
2. on exhausting 2, keep increasing the divisor by 1 - the next divisor which completely divides has to be the next prime
3. continue doing this till the number is not left

In [7]:
number = 600851475143
prime_list = [2]
cur_prime = 2

while number != 1:
    while number % cur_prime == 0:
        number = number / cur_prime
        if prime_list[-1] != cur_prime:
            prime_list.append(cur_prime)
        
    cur_prime += 1
    
print "Answer is", cur_prime-1
print prime_list

Answer is 6857
[2, 71, 839, 1471, 6857]


Okay. This works as expected. Nothing facy here.
While unlikely, it's always possible that `number` is itself a prime. Not sure if I can avoid couting upto that number though - but counting is fast, right?

In [9]:
_c = 1
while _c<=10000000:
    _c += 1

No. Counting is slow. Counting upto `10 million` is actually pretty slow. What I can do, is eliminate all factors of 2 in the beginning. After that I can set `cur_prime` as 3 and increase by 2 - this should cut down the counting by half.

In [28]:
number = 99
two = 2
cur_prime = 3

## check with 2
while number % two == 0:
    number = number / two
    
if number == 1:
    print "Answer is 2"
    
## now for whatever's left, you can increase the counter by 2 and cut the counting by half
answer = 3
while number != 1:
    while number % cur_prime == 0:
        answer = cur_prime
        number = number / cur_prime
    
    cur_prime += 2

print "Answer is", answer

Answer is 11


Okay, so there is indeed a better solution to the counting problem explained on Euler.

The idea is that any number can have atmost one prime above it's square root.
- Start from 2.
- Eliminate all factors of 2.
- Take what's left and compute it's square root.
    - If the next candidate (i.e. 3) is greater than the current square root, then that's the answer.
    - Else factorize with current candidate and repeat the process.

In [23]:
number = 103

## check for 2s
while number%2 == 0:
    number = number / 2
    
if number == 1:
    print "Answer is 2"
        
else:
    f = 3
    limit = math.sqrt(number)
    while number > 1 and f <= limit:
        if number % f == 0:
            number = number / f
            answer = f
            print "ANSWER:", answer
            ## exhaustively divide the number.
            while number % f == 0:
                number = number / f
                
            limit = math.sqrt(number)
        
            ## move to next f
        f += 2
            
    if number == 1:
        print answer
    else:
        print number, limit, f

103 10.1488915651 11


## Problem 4

A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.

In [49]:
111 ** 2

12321

Making the palindrome would involve 111.

In [82]:
%%timeit
largest = -1
for i in range(100, 999):
    for j in range(100, 999):
        n = i * j
        snum = str(n)
        if snum == snum[::-1]:
            if n > largest:
                largest = n
                _one = i
                _two = j
            
print largest, i, _one, _two

906609 998 913 993
906609 998 913 993
906609 998 913 993
906609 998 913 993
1 loop, best of 3: 644 ms per loop


In [75]:
largest = 12321
for i in range(12321, 999*999, 111):
    if i % 111 == 0
    else:
        snum = str(i)
        if snum == snum[::-1]:
            if i > largest:
                largest = i
            
print "Largest", largest

Largest 987789


In [69]:
snum[::-1]

'400699'

Unfortunately, this is not correct. While it will give me the largest palindrome, it might not be a product of 2 three digit numbers. My best solution is to brute force while first checking for divisibility with 11 - only the str and reverse function takes time.

In [81]:
%%timeit
largest = -1
for i in xrange(100, 1000):
    for j in xrange(99, 1000, 11):
        n = i*j
        snum = str(n)
        if snum == snum[::-1]:
            if n>largest:
                largest = n
            
print largest

906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
10 loops, best of 3: 60.2 ms per loop


Yup, this is 10 times faster: from 644ms to 60.2ms

Now, I looked at Euler. There first solution is brute force but not with string reverse.
Maybe the string transformation is taking time. 

In [90]:
def find_reverse(n):
    ## keep dividing by 10 as long as the number remains.
    rnum = 0
    while n > 0:
        d = n % 10
        rnum = rnum * 10 + d
        n = n / 10
        
    return rnum

In [91]:
%%timeit
largest = -1
for i in range(100, 1000):
    for j in range(99, 1000, 11):
        n = i * j
        if n > largest:
            if n == find_reverse(n):
                largest = n
                
print largest

906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
10 loops, best of 3: 35.7 ms per loop


:) Ditching the string reverse check cuts down the time (almost)by half again. From 644ms to 60.2ms to 35.2ms .

They also point out that I will inevitably be doing repeated computations because a and b will keep switching values. Perhaps I can loop beginning from 

In [98]:
%%timeit
largest = -1
for i in range(100, 1000):
    for j in range(99, 1000, 11):
        if i>=j:
            n = i * j
            if n > largest:
                if n == find_reverse(n):
                    largest = n
                
print largest

906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
906609
10 loops, best of 3: 30.3 ms per loop


Yup, this helps. Cuts down another 5ms. Other "hhACKZZzzZsSS" include counting backward from 999 to encounter bigger numbers first.

## Problem 5

2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

In [133]:
def find_remaining_factors(n, f):
    remainder = 1
    ## eliminate 2
    while f%2 == 0:
        f = f / 2
        if n%2 == 0:
            n = n / 2
        else:
            remainder *= 2

    if f == 1:
        return remainder

    else:
        p = 3
        while f > 1:
            while f%p == 0:
                f = f / p
                if n%p == 0:
                    n = n / p
                else:
                    remainder *= p
            p += 2

        return remainder

In [156]:
find_remaining_factors(5, 3)

3

In [166]:
factors = range(1, 11)
n = 1

for f in factors:
    if n%f == 0:
        ## nothing to do here...
        pass
    else:
        ## multiply n by the prime-factors in f which are not in n
        n = n * find_remaining_factors(n, f)

print n

2520


So, the code above replicates the example. To find evenly divisible numbers, we ensure that it divides completely and by an even number... or, I can just modify the list so that it becomes 2, 4, 6, 8,... instead of 1, 2, 3, 4... .

In [157]:
def EVEN_find_remaining_factors(n, f):
    remainder = 1
    ## eliminate 2
    while f%2 == 0:
        f = f / 2
        if n%2 == 0:
            n = n / 2
        else:
            remainder *= 2
    
    if f == 1:
        if remainder%2 != 0:
            remainder *= 2
        return remainder
    
    else:
        p = 3
        while f > 1:
            while f%p == 0:
                f = f / p
                if n%p == 0:
                    n = n / p
                else:
                    remainder *= p
            p += 2
            
        if remainder%2 != 0:
            remainder *= 2
        return remainder

In [176]:
factors = range(1, 21)
n = 2

for f in factors:
    if n%f == 0:
        ## nothing to do here...
        #if (n/f)%2 != 0:
            #n = n * 2
        pass
    else:
        ## multiply n by the prime-factors in f which are not in n
        n = n * find_remaining_factors(n, f)
        ## you can check for even now...
        #if (n/f)%2 != 0:
            #n = n * 2

print n

for f in factors:
    print f, n/f
    if (n/f)%2 != 0:
        n = n * 2
        
print n

232792560
1 232792560
2 116396280
3 77597520
4 58198140
5 46558512
6 38798760
7 33256080
8 29099070
9 25865840
10 23279256
11 21162960
12 19399380
13 17907120
14 16628040
15 15519504
16 14549535
17 27387360
18 25865840
19 24504480
20 23279256
465585120


The previously accepted answer of 232792560 is not valid since 16 does not evenly divide this number.

## Problem 6

The sum of the squares of the first ten natural numbers is,

12 + 22 + ... + 102 = 385
The square of the sum of the first ten natural numbers is,

(1 + 2 + ... + 10)2 = 552 = 3025
Hence the difference between the sum of the squares of the first ten natural numbers and the square of the sum is 3025 − 385 = 2640.

Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.

In [178]:
numbers = range(1, 101)

sum_squares, squares_sum = 0, 0

for n in numbers:
    sum_squares = sum_squares + n**2
    squares_sum += n
    
print squares_sum**2 - sum_squares

25164150


## Problem 7

By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.

What is the 10 001st prime number?

In [187]:
%%timeit
ix = 1
primes = [2]
counter = 3
while ix < 10001:
    ## 2s
    isPrime = True
    for p in primes:
        if counter%p == 0:
            isPrime = False
            break
            
    if isPrime:
        ix += 1
        primes.append(counter)
        
    counter += 2
    
print primes[-1]

104743
104743
104743
104743
1 loop, best of 3: 8.71 s per loop


The code above is the naive approach which takes . And I can't think of anything better. Off to Euler's solution page.

Okay. The very first hint is the form of primes. If I start with [2, 3], then every prime after that can be of form `6k +- 1`. So, I can jump in steps of 6 and check for +1 and -1.

In [195]:
def stupid_check_prime(n, primes):
    isPrime = True

    for p in primes:
        if n%p == 0:
            isPrime = False
            break
            
    return isPrime

In [197]:
%%timeit
primes = [2, 3]
ix = len(primes)
counter = 6
while ix < 10001:
    isPrime = stupid_check_prime(counter - 1, primes)
    if isPrime:
        ix += 1
        primes.append(counter - 1)
        
    
    isPrime = stupid_check_prime(counter + 1, primes)
    if isPrime:
        ix += 1
        primes.append(counter + 1)
        
    counter += 6
    
print primes[-1]

104743
104743
104743
104743
1 loop, best of 3: 8.69 s per loop


Unfortunately, the gain is minimal... probably because for every jump of 6, I am check for 2 values, whereas previously, I was checking 1 value in 3 jumps... So this is not helpful at all. 

But there is that square-root procedure which I can use : (Sieve Sieve) they keep calling it.

In [8]:
%%timeit
ix = 1
primes = [2]
counter = 3
while ix < 10001:
    ## 2s
    isPrime = True
    limit = math.sqrt(counter)
    p_ix = 0
    while (primes[p_ix]<=limit) and p_ix<(ix):
        if counter%primes[p_ix] == 0:
            isPrime = False
            break
        p_ix += 1
            
    if isPrime:
        ix += 1
        primes.append(counter)
        
    counter += 2
    
print primes[-1]

104743
104743
104743
104743
1 loop, best of 3: 420 ms per loop


Wow. This decreased significantly. From 8.7s to 425ms . Just for thoroughness, I will re-write the max-prime problem using this.

In [36]:
n = 23
p = 2
while n%p == 0:
    n = n/p
    
if n == 1:
    print "Answer", p
    
else:
    p = 3
    limit = math.sqrt(n)
    while (p <= limit) and (n>1):
        print p
        while n%p == 0:
            n = n/p
        
        limit = math.sqrt(n)
        cur_prime = p
        p += 2
    
    if n == 1:
        print "Answer", cur_prime
    else:
        print "Answer", n

3
Answer 23


This cuts down the steps even more, since I only have to iterate upto the `math.sqrt(n)`.

## Problem 8

The four adjacent digits in the 1000-digit number that have the greatest product are 9 × 9 × 8 × 9 = 5832.

```73167176531330624919225119674426574742355349194934
96983520312774506326239578318016984801869478851843
85861560789112949495459501737958331952853208805511
12540698747158523863050715693290963295227443043557
66896648950445244523161731856403098711121722383113
62229893423380308135336276614282806444486645238749
30358907296290491560440772390713810515859307960866
70172427121883998797908792274921901699720888093776
65727333001053367881220235421809751254540594752243
52584907711670556013604839586446706324415722155397
53697817977846174064955149290862569321978468622482
83972241375657056057490261407972968652414535100474
82166370484403199890008895243450658541227588666881
16427171479924442928230863465674813919123162824586
17866458359124566529476545682848912883142607690042
24219022671055626321111109370544217506941658960408
07198403850962455444362981230987879927244284909188
84580156166097919133875499200524063689912560717606
05886116467109405077541002256983155200055935729725
71636269561882670428252483600823257530420752963450```

Find the thirteen adjacent digits in the 1000-digit number that have the greatest product. What is the value of this product?

In [1]:
sequence = """96983520312774506326239578318016984801869478851843
85861560789112949495459501737958331952853208805511
12540698747158523863050715693290963295227443043557
66896648950445244523161731856403098711121722383113
62229893423380308135336276614282806444486645238749
30358907296290491560440772390713810515859307960866
70172427121883998797908792274921901699720888093776
65727333001053367881220235421809751254540594752243
52584907711670556013604839586446706324415722155397
53697817977846174064955149290862569321978468622482
83972241375657056057490261407972968652414535100474
82166370484403199890008895243450658541227588666881
16427171479924442928230863465674813919123162824586
17866458359124566529476545682848912883142607690042
24219022671055626321111109370544217506941658960408
07198403850962455444362981230987879927244284909188
84580156166097919133875499200524063689912560717606
05886116467109405077541002256983155200055935729725
71636269561882670428252483600823257530420752963450"""

In [2]:
unrolled = sequence.strip().replace('\n', '')
print len(unrolled)

950


In [6]:
unrolled[-13:]

'0420752963450'

In [12]:
maxsubs = -1
for i in range(0, len(unrolled)-13):
    subs = unrolled[i:i+13]
    if '0' not in subs:
        p = 1
        for n in subs:
            p *= int(n)
        
        if p > maxsubs:
            maxsubs = p
print "Answer", maxsubs

Answer 23514624000


The simplest way would be to iterate from the first position till the last possible position (i.e. after which you still have 13 digits). If there is a 0 in the subsequence then skip else compute the product of digits.

## Problem 9

A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,

a2 + b2 = c2

For example, 3^2 + 4^2 = 9 + 16 = 25 = 5^2.

There exists exactly one Pythagorean triplet for which a + b + c = 1000.
Find the product abc.

**solution** : tried brute forcing everything upto 1000. Then checked online and found about Euclids formula for finding triplets. Good stuff. Used it to derive the solution analytically.

## Problem 10

The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

Find the sum of all the primes below two million.

Okay. This seems easy enough. I'll just have to recursively apply the priming condition on every number, while also maintaining a memoized lookup of other factors.

In [1]:
import math

In [9]:
%%timeit
primes = [2]
for n in xrange(3, 2000001):
    res = check_prime(n, primes)
    if res:
        primes.append(n)
        
print sum(primes)

142913828922
142913828922
142913828922
142913828922
1 loop, best of 3: 20.1 s per loop


In [4]:
def check_prime(number, prime_list):
    limit = math.sqrt(number)
    if number % 2 == 0:
        return False
    else:
        pix = 1
        while (pix < len(prime_list)) and (prime_list[pix] <= limit):
            factor = prime_list[pix]
            if number % factor == 0:
                return False
            else:
                pix += 1
                
        return True

In [48]:
def erether_prime(n):
    fc = 0
    factors = []
    if n%2 == 0:
        return False
        factors.append(2)
        fc += 1
        while n%2 == 0:
            n = n/2
            
    else:
        start = 3
        limit = math.sqrt(n)
        
        while start <= limit:
            if n%start == 0:
                return False
                factors.append(start)
                fc += 1
                while n%start == 0:
                    n = n/start
                    
                limit = math.sqrt(n)
                start += 2
            else:
                start += 2
                
    if fc == 0:
        return True
    else:
        return False

In [49]:
%%timeit
_sprime = 0
_lprime = 2
for n in xrange(3, 2000001):
    if erether_prime(n):
        _lprime = n
        _sprime += _lprime
        
print _lprime, _sprime

1999993 142913828920
1999993 142913828920
1999993 142913828920
1999993 142913828920
1 loop, best of 3: 30.9 s per loop


In [38]:
number = 97
start = 3
prime_factors = []

while number%2 == 0:
    number = number / 2

if number == 1:
    print "Largest prime:", 2
else:
    limit = math.sqrt(number)
    while start <= limit:
        if number % start == 0:
            cur_prime = start
            prime_factors.append(cur_prime)
            while number % start == 0:
                number = number/start

        else:
            start += 2
            limit = math.sqrt(number)

    if number == 1:
        print "Largest prime:", cur_prime
    else:
        print "Largest prime:", number
        
print prime_factors

Largest prime: 97
[]


### Finally, trying the actual sieve method.

In [50]:
len(range(3, 2000001, 2))

999999

In [ ]:
ps = 2
n = 2000000
sieve = range(3, 2000001, 2)
limit = math.sqrt(n)

ix = 0

while ix<=999999:
    if sieve[ix] > limit:
        ix += 1
    else:
        

# scratch